### PR TITLE
Fix serialization errors for Dispatcher.jl

### PR DESCRIPTION
**Pull request by nicoleepp**
This is a fix for issue invenia/Dispatcher.jl#21

A new version of DeferredFutures should probably be released once this has merged so that Dispatcher.jl can reference that and avoid build failures.

I also got rid of the deprecated warnings that show up on Julia 0.6

### CODE REVIEW

File: src/DeferredFutures.jl [hunk 1]
```diff
@@ -53,6 +54,8 @@ end
 Serialize a DeferredFuture such that it can de deserialized by `deserialize` in a cluster.
 """
 function Base.serialize(s::AbstractSerializer, ref::DeferredFuture)
+    Base.Serializer.serialize_cycle(s, ref) && return
```
Comment (rofinn) [inline src/DeferredFutures.jl:57]: Maybe there should be a simplified test which checks that the `serialize_cycle(s, ref) && return` solves the issue discovered in Dispatcher?

Comment (nicoleepp) [reply src/DeferredFutures.jl:57]: I'm not sure how to test that without adding Dispatcher.jl as a dependency for tests, which might be a good idea. I'll try that tomorrow.

**Comment by rofinn** [comment]
Looks like the mac 0.5 build hit the HDF5 issue. [PR 430](https://github.com/JuliaIO/HDF5.jl/pull/430) should fix it once it gets merged and a new release is tagged.

File: test/REQUIRE [hunk 2]
```diff
@@ -0,0 +1 @@
+Dispatcher
```
Comment (rofinn) [inline test/REQUIRE]: Might want to specify a minimum version number.

File: test/runtests.jl [hunk 3]
```diff
@@ -477,5 +478,22 @@ using Compat
...
+            @everywhere using Dispatcher
+
+            try
+                ctx = DispatchContext()
+                exec = ParallelExecutor()
+
+                op = Op(()->3)
+                add!(ctx, op)
+
+                results = run!(exec, ctx)
```
Comment (rofinn) [inline test/runtests.jl]: Should probably test that this actually returns 3 (vs a wrapped exception or something). Also, I'm a little worried about the DeferredFuture tests depending on dispatcher. Couldn't we just trigger this behaviour by passing a nested `DeferredFuture` between two processes (using `@spawn` directly)?

**Review by rofinn — CHANGES_REQUESTED** [review]

File: test/runtests.jl (hunk 3 above)
Comment (nicoleepp) [reply test/runtests.jl]: You mean like this?

```
using DeferredFutures
addprocs(2)
@everywhere using DeferredFutures

df = DeferredFuture()

@spawn begin
    df2 = DeferredFuture()
    put!(df2, 3)
    put!(df, df2)
end

@assert fetch(fetch(df)) == 3
```

This seems to work fine. I've tried recreating the error without using Dispatcher but haven't been able to yet.

Comment (rofinn) [reply test/runtests.jl]: Hmmm, yeah, I'm not entirely sure either. Did a bit of digging though and we could probably adapt the base julia tests for serialization cycles https://github.com/JuliaLang/julia/blob/master/test/serialize.jl#L350 ?

Comment (nicoleepp) [reply test/runtests.jl]: Ok I will look into doing that

File: src/DeferredFutures.jl (hunk 1 above)
Comment (nicoleepp) [reply src/DeferredFutures.jl:57]: Finally figured out how to recreate the error that manifests itself in Dispatcher.jl. Tests have been updated.

Comment (rofinn) [reply src/DeferredFutures.jl:57]: Awesome, thank you.

File: test/runtests.jl [hunk 4]
```diff
@@ -477,5 +477,37 @@ using Compat
             @test_broken isready(bottom_dc3) == true
             @test_broken fetch(bottom_dc3) == 14
         end
+
+        @testset "DeferredFuture serialization as part of another object" begin
```
Comment (rofinn) [inline test/runtests.jl:481]: Yeah, these tests look a lot cleaner now. It's also nice that it clearly documents the old failure condition outside of dispatcher.

**Review by rofinn — APPROVED** [review]
Nice work!

**Comment by iamed2** [comment]
Great work!

**Pull request merged**